enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,327 @@
+
+The external CCM will read its configuration from `kube-cloud-config-external`.
+
+### Preventing upgrade when config cannot be upgraded
+
+If CCO is not able to write `kube-cloud-config-external` it will mark itself as
+non-upgradable with a user-actionable error message explaining how
+`openshift-config/cloud-provider-config` is incompatible.
+
+As we intend to switch to external CCM in 4.11, this change will be implemented
+in OCP 4.10.
+
+### Compatibility constraints on user-managed config
+
+This has implications for the user-managed configuration:
+
+In 4.10, the user-managed configuration **must** be compatible with the legacy
+cloud provider. The user **must not** specify config options which cannot be
+transformed to config for the external cloud provider. If either of these are
+not satisfied we will mark CCO as not upgradeable.
+
+In 4.11, the base `cloud-provider-config` **may** include config options which
+are not compatible with the legacy cloud provider. If these options cannot be
+transformed for the legacy cloud provider we will not update
+`kube-cloud-config`, but will not delete the existing one.
+
+### User Stories
+
+#### Story 1 - Fresh install
+
+The Tyrell Corporation wants to deploy OCP 4.11 on their OSP cloud. OCP
+automatically configures external CCM out-of-the-box with configuration native
+to external OpenStack CCM.
+
+#### Story 2 - Upgrade with default configuration
+
+The Wallace Corporation wants to upgrade their OCP 4.10 cluster to OCP 4.11.
+They are using the default cloud provider configuration. Upon upgrading, their
+cluster should automatically use CCM following the upgrade, and the upgrade
+should require no additional user input.
+
+#### Story 3 - Upgrade with non-default compatible configuration
+
+The Acme Corporation wants to upgrade their OCP 4.10 cluster to OCP 4.11.
+Following the user documentation [here][cloud-provider-options], they have
+configured a specific floating network ID. Upon upgrading, their cluster should
+automatically use CCM following the upgrade, and the upgrade should require no
+user input for this aspect of the upgrade. After the upgrade, their cluster
+should continue to use the same floating network ID.
+
+[cloud-provider-options]: https://docs.openshift.com/container-platform/4.9/installing/installing_openstack/installing-openstack-installer-custom.html#installation-osp-setting-cloud-provider-options_installing-openstack-installer-custom
+
+#### Story 4 - Upgrade with non-default incompatible configuration
+
+Omni Consumer Products wants to upgrade their OCP 4.10 cluster to OCP 4.11.
+They have configured a separate secret containing their cloud credentials as
+well as additional config directives. Upon attempting to upgrade, they discover
+they cannot as the Cluster Config Operator is marked as non-upgradable. The
+status of this operator informs them of the incompatible config directives and
+provides guidance on how to remedy the situation.
```

Comment:
A concrete example of a scenario of invalid config and the remediation steps would be useful here